lib/srv/discovery/access_graph.go

```diff
@@ -195,15 +197,49 @@ func newAccessGraphClient(ctx context.Context, certs []tls.Certificate, config s
 // watches the connection state. If the connection is closed, it will
 // automatically try to reconnect.
 func (s *Server) initializeAndWatchAccessGraph(ctx context.Context, reloadCh <-chan struct{}) error {
-	// Configure health check service to monitor access graph service and
-	// automatically reconnect if the connection is lost without
-	// relying on new events from the auth server to trigger a reconnect.
-	const serviceConfig = `{
+	const (
+		// aws discovery semaphore lock.
+		semaphoreName = "aws_discovery"
+		// Configure health check service to monitor access graph service and
+		// automatically reconnect if the connection is lost without
+		// relying on new events from the auth server to trigger a reconnect.
+		serviceConfig = `{
 		 "loadBalancingPolicy": "round_robin",
 		 "healthCheckConfig": {
 			 "serviceName": ""
 		 }
 	 }`
+	)
+	// AcquireSemaphoreWithRetry will retry until the semaphore is acquired.
+	// This prevents multiple discovery services to push AWS resources in parallel.
+	const (
+		semaphoreExpiration = time.Minute
+	)
+
+	lease, err := services.AcquireSemaphoreLock(
```

Comment:
In the comment above we talk about `AcquireSemaphoreWithRetry` but we're actually using `AcquireSemaphoreLock`.